Failure to setup client after installation

**Describe the bug**
After installing Fex on my arm64 laptop (Lenovo Yoga Slim7x) running Ubuntu Concept 24.10, choosing the Ubuntu 24.04 SquashFS, using it as-is and setting it as the default, the "trying basic program run" step fails with the following error message:

> FEX is now installed. Trying basic program run
> [ERROR] Couldn't connect to FEXServer socket 1000.FEXServer.Socket 111
> [ERROR] Couldn't connect to FEXServer socket /run/user/1000/1000.FEXServer.Socket 111
> [ERROR] Couldn't connect to FEXServer socket 1000.FEXServer.Socket 111
> [ERROR] Couldn't connect to FEXServer socket /run/user/1000/1000.FEXServer.Socket 111
> [ERROR] Couldn't connect to FEXServer socket 1000.FEXServer.Socket 111
> [ERROR] Couldn't connect to FEXServer socket /run/user/1000/1000.FEXServer.Socket 111
> [ERROR] Couldn't connect to FEXServer socket 1000.FEXServer.Socket 111
> [ERROR] Couldn't connect to FEXServer socket /run/user/1000/1000.FEXServer.Socket 111
> [ERROR] Couldn't connect to FEXServer socket 1000.FEXServer.Socket 111
> [ERROR] Couldn't connect to FEXServer socket /run/user/1000/1000.FEXServer.Socket 111
> [ERROR] Couldn't connect to FEXServer socket 1000.FEXServer.Socket after launching the process
> [ERROR] FEXServerClient: Failure to setup client
> FEXInterpreter failed to run. Not continuing
   

**To Reproduce**
Steps to reproduce the behavior:
1. Have a Lenovo Yoga Slim7x laptop running Ubuntu Concept 24.10.
2. Run the installer with `curl --silent https://raw.githubusercontent.com/FEX-Emu/FEX/main/Scripts/InstallFEX.py --output /tmp/InstallFEX.py && python3 /tmp/InstallFEX.py && rm /tmp/InstallFEX.py`
3. Select `Ubuntu 24.04 (SquashFS)` in the RootFS selection step.
4. Use the squashFS file as-is (option 2).
5. Set Ubuntu_24_04.sqsh as the default RootFS.
6. Installer crashes with above error.

**Expected behavior**
Expected basic program run to succeed.

**Screenshots and Video**
N/A.

**System information:**
 - OS: Ubuntu Concept 24.10
 - CPU/SoC: Snapdragon X1 Elite
 - Video driver version: N/A (yes, really! no Adreno X1-85 gfx driver released for linux yet)
 - RootFS used: Ubuntu 24.04 (SquashFS)
 - FEX version: FEX-2501
 - Thunks Enabled: Don't know.

**Additional context**
 - Is this an x86 or x86-64 game: N/A
 - Does this reproduce on AArch64 with Radeon/Intel/Nvidia: Untested
 - Is this a Vulkan game: No

Add any other context about the problem here.

> $ lsb_release -a
> No LSB modules are available.
> Distributor ID:	Ubuntu
> Description:	Ubuntu 24.10
> Release:	24.10
> Codename:	oracular

> $ uname -a
> Linux ola-Yoga-Slim-7-14Q8X9 6.12.0-28-qcom-x1e #28-Ubuntu SMP PREEMPT_DYNAMIC Fri Jan 24 03:56:34 UTC 2025 aarch64 aarch64 aarch64 GNU/Linux

--- a/Source/Common/FEXServerClient.cpp
+++ b/Source/Common/FEXServerClient.cpp
@@ -270,10 +270,13 @@ int ConnectToAndStartServer(char* InterpreterPath) {
       // Child
       close(fds[0]); // Close read end of pipe
 
-      const char* argv[2];
+      const char* argv[4];
 
+      auto pipe_string = fextl::fmt::format("{}", fds[1]);
       argv[0] = FEXServerPath.c_str();
-      argv[1] = nullptr;
+      argv[1] = "--wait_pipe";
+      argv[2] = pipe_string.c_str();
+      argv[3] = nullptr;
 
       if (execvp(argv[0], (char* const*)argv) == -1) {
         // Let the parent know that we couldn't execute for some reason

--- a/Source/Tools/FEXServer/ArgumentLoader.cpp
+++ b/Source/Tools/FEXServer/ArgumentLoader.cpp
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: MIT
 #include "ArgumentLoader.h"
 #include "Common/cpp-optparse/OptionParser.h"
+#include "PipeScanner.h"
 
 #include "git_version.h"
 
@@ -20,6 +21,7 @@ FEXServerOptions Load(int argc, char** argv) {
   Parser.add_option("-p", "--persistent").action("store").type("int").set_default(0).set_optional_value(true).metavar("n").help("Make FEXServer persistent. Optional number of seconds");
 
   Parser.add_option("-w", "--wait").action("store_true").set_default(false).help("Wait for the FEXServer to shutdown");
+  Parser.add_option("--wait_pipe").action("store").type("int").set_default(-1).set_optional_value(true);
 
   Parser.add_option("-v").action("version").help("Version string");
 
@@ -33,6 +35,10 @@ FEXServerOptions Load(int argc, char** argv) {
   }
   FEXOptions.PersistentTimeout = Options.get("persistent");
 
+  int WaitPipe = Options.get("wait_pipe");
+  if (WaitPipe != -1) {
+    PipeScanner::SetWaitPipe(WaitPipe);
+  }
   return FEXOptions;
 }
 } // namespace FEXServer::Config

--- a/Source/Tools/FEXServer/Main.cpp
+++ b/Source/Tools/FEXServer/Main.cpp
@@ -109,10 +109,6 @@ int main(int argc, char** argv, char** const envp) {
     LogMan::Msg::InstallHandler(Logging::MsgHandler);
   }
 
-  // Scan for any incoming pipes
-  // We will close these later
-  PipeScanner::ScanForPipes();
-
   if (!Options.Foreground) {
     DeparentSelf();
   }

--- a/Source/Tools/FEXServer/PipeScanner.cpp
+++ b/Source/Tools/FEXServer/PipeScanner.cpp
@@ -1,42 +1,18 @@
 // SPDX-License-Identifier: MIT
-#include <cstdlib>
 #include <dirent.h>
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <unistd.h>
 #include <vector>
+#include <fcntl.h>
 
 namespace PipeScanner {
 std::vector<int> IncomingPipes {};
-
-// Scan and store any pipe files.
-// This will capture all pipe files so needs to be executed early.
-// This ensures we find any pipe files from execve for waiting FEXInterpreters.
-void ScanForPipes() {
-  DIR* fd = opendir("/proc/self/fd");
-  if (fd) {
-    struct dirent* dir {};
-    do {
-      dir = readdir(fd);
-      if (dir) {
-        char* end {};
-        int open_fd = std::strtol(dir->d_name, &end, 8);
-        if (end != dir->d_name) {
-          struct stat stat {};
-          int result = fstat(open_fd, &stat);
-          if (result == -1) {
-            continue;
-          }
-          if (stat.st_mode & S_IFIFO) {
-            // Close any incoming pipes
-            IncomingPipes.emplace_back(open_fd);
-          }
-        }
-      }
-    } while (dir);
-
-    closedir(fd);
-  }
+void SetWaitPipe(int FD) {
+  int flags = fcntl(FD, F_GETFD);
+  flags |= FD_CLOEXEC;
+  fcntl(FD, F_SETFD, flags);
+  IncomingPipes.emplace_back(FD);
 }
 
 void ClosePipes() {

--- a/Source/Tools/FEXServer/PipeScanner.h
+++ b/Source/Tools/FEXServer/PipeScanner.h
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: MIT
 #pragma once
 namespace PipeScanner {
-void ScanForPipes();
+void SetWaitPipe(int FD);
 void ClosePipes();
 } // namespace PipeScanner

